File: 139.java
// Leetcode 139 Word Break
// Reference: https://leetcode.com/problems/word-break/solutions/44054/java-dp-solution/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

import java.util.List;

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        if (s == null || s.length() == 0) return false;

        int n = s.length();

        // dp[i] represents whether s[0...i] can be formed by dict
        boolean[] dp = new boolean[n];

        for(int i = 0; i < n; i++){
            for(int j = 0; j <= i; j++){
                String sub = s.substring(j, i+1);

                if(wordDict.contains(sub) && (j == 0 || dp[j-1]){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n-1];
    }
}

File: 139.js
// Leetcode 139 Word Break
// Reference: https://leetcode.com/problems/word-break/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    if(wordDict == null || wordDict.length === 0) return false;

    const set = new Set(wordDict);
    const dp = Array(s.length + 1).fill(false);
    dp[0] = true;

    for(let end = 1; end <= s.length; end++){
        for (let start = 0; start < end; start ++){
            const w = s.slice(start, end);
            if (dp[start] === true && set.has(w)){
                dp[end] = true;
                break;
            }
        }
    }
    return dp[s.length];
};